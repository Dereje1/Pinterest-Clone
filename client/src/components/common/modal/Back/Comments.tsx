import React from 'react';
import CardContent from '@mui/material/CardContent';
import Typography from '@mui/material/Typography';
import Fab from '@mui/material/Fab';
import AddIcon from '@mui/icons-material/Add';
import Card from '@mui/material/Card';
import StarBorderIcon from '@mui/icons-material/StarBorder';
import CloseIcon from '@mui/icons-material/Close';
import Avatar from '@mui/material/Avatar';
import CommentForm from './CommentForm';
import PinnersDialog from './PinnersDialog';
import { ProfileLink } from '../../common';
import Tags from './Tags';
import { formatDate } from '../../../../utils/utils';
import { PinType } from '../../../../interfaces';

interface CommentsProps {
  stylingProps: {width: number, height: number}
  pinInformation: PinType
  authenticated: boolean
  handleNewComment: (comment: string) => void
  toggleComments: () => void
  closePin: (_: React.SyntheticEvent, forceClose?: boolean) => void
  updateTags: (query: string) => void
  displayLogin: () => void
}

function Comments({
  stylingProps,
  pinInformation,
  authenticated,
  handleNewComment,
  toggleComments,
  closePin,
  updateTags,
  displayLogin,
}: CommentsProps) {
  const [openCommentForm, setOpenCommentForm] = React.useState(false);
  const [openPinnersDialog, setOpenPinnersDialog] = React.useState(false);

  const handleOpenCommentForm = () => {
    setOpenCommentForm(true);
  };

  const handleCloseCommentForm = () => {
    setOpenCommentForm(false);
  };

  const submitComment = (comment: string) => {
    if (comment.trim().length) {
      handleNewComment(comment);
    }
    setOpenCommentForm(false);
  };

  return (
    <div style={{ ...stylingProps, overflowY: 'auto', paddingBottom: 3 }}>
      <div style={{
        position: 'absolute',
        bottom: 10,
        left: 0,
        right: 0,
        width: 300,
        display: 'flex',
        justifyContent: 'space-around',
        marginLeft: 'auto',
        marginRight: 'auto',
      }}
      >
        <Fab
          color="error"
          aria-label="close"
          onClick={closePin}
        >
          <CloseIcon fontSize="large" />
        </Fab>

        <Fab
          color="success"
          aria-label="pinners"
          onClick={() => setOpenPinnersDialog(true)}
          disabled={!pinInformation.savedBy.length}
        >
          <StarBorderIcon fontSize="large" />
        </Fab>

        <Fab
          color="primary"
          aria-label="add"
          onClick={handleOpenCommentForm}
          disabled={!authenticated || openCommentForm}
        >
          <AddIcon fontSize="large" />
        </Fab>
      </div>

      <div style={{
        display: 'flex',
        justifyContent: 'space-between',
        alignItems: 'center',
        position: 'sticky',
        top: 0,
        background: 'white',
        marginBottom: 10,
      }}
      >
        <div
          style={{
            display: 'flex',
            justifyContent: 'flex-start',
            alignItems: 'center',
            width: '100%',
            paddingLeft: 10,
            paddingRight: 10,
          }}
        >

          <Avatar
            id="thumbnail"
            alt={pinInformation.imgDescription}
            src={pinInformation.imgLink}
            style={{ width: 56, height: 56, cursor: 'pointer' }}
            onClick={toggleComments}
          />

          { !openCommentForm && (
            <Tags
              commentFormIsOpen={openCommentForm}
              updateTags={updateTags}
              pinInformation={pinInformation}
              closePin={closePin}
            />
          )}
        </div>
      </div>
      {
        openCommentForm && (
          <CommentForm
            handleClose={handleCloseCommentForm}
            handleSubmit={submitComment}
          />
        )
      }
      {
        !openCommentForm && !pinInformation.comments.length && (
          <Card
            sx={{ margin: 1 }}
            raised
          >
            <CardContent>
              <Typography variant="h6" color="text.secondary" sx={{ margin: 'auto' }}>
                {authenticated
                  ? `Be the first to write a comment on ${pinInformation.imgDescription}...`
                  : `Login and be the first to write a comment on ${pinInformation.imgDescription}...`}
              </Typography>
            </CardContent>
          </Card>
        )
      }
      {
        !openCommentForm && pinInformation.comments.map(({
          _id, comment, displayName, createdAt, userId,
        }, idx) => (
          <React.Fragment key={_id}>
            <Card
              sx={{ margin: 1 }}
              raised
            >
              <CardContent>
                <div style={{ display: 'flex', alignItems: 'center' }}>
                  <ProfileLink
                    authenticated={authenticated}
                    closePin={closePin}
                    displayLogin={displayLogin}
                    title={(
                      <Typography variant="subtitle1" sx={{ color: '#3752ff', fontWeight: 'bold' }}>
                        {`${displayName}`}
                      </Typography>
                    )}
                    userId={userId}
                  />
                  <Typography color="text.secondary" sx={{ ml: 1, fontSize: 12 }}>
                    {`- ${formatDate(createdAt)}`}
                  </Typography>
                </div>
                <Typography sx={{ mb: 1.5 }} color="text.secondary">
                  {comment}
                </Typography>
              </CardContent>
            </Card>
            {idx === pinInformation.comments.length - 1 && <div style={{ width: '100%', height: 65 }} />}
          </React.Fragment>
        ))
      }
      <PinnersDialog
        authenticated={authenticated}
        displayLogin={displayLogin}
        open={openPinnersDialog}
        onCloseDialog={() => setOpenPinnersDialog(false)}
        pinnersList={pinInformation.savedBy}
        onClosePin={closePin}
      />
    </div>
  );
}

export default Comments;
