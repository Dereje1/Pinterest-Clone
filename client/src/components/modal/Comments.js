import React from 'react';
import PropTypes from 'prop-types';
import CardContent from '@mui/material/CardContent';
import Typography from '@mui/material/Typography';
import Fab from '@mui/material/Fab';
import AddIcon from '@mui/icons-material/Add';
import Card from '@mui/material/Card';
import StarBorderIcon from '@mui/icons-material/StarBorder';
import CloseIcon from '@mui/icons-material/Close';
import Avatar from '@mui/material/Avatar';
import CommentForm from './CommentForm';
import PinnersDialog from './PinnersDialog';
import Tags from './Tags';
import { formatDate } from '../../utils/utils';

function Comments({
  stylingProps,
  pinInformation,
  handleNewComment,
  authenticated,
  toggleComments,
  closePin,
  updateTags,
}) {
  const [openCommentForm, setOpenCommentForm] = React.useState(false);
  const [openPinnersDialog, setOpenPinnersDialog] = React.useState(false);

  const handleOpenCommentForm = () => {
    setOpenCommentForm(true);
  };

  const handleCloseCommentForm = () => {
    setOpenCommentForm(false);
  };

  const submitComment = (comment) => {
    if (comment.trim().length) {
      handleNewComment(comment);
    }
    setOpenCommentForm(false);
  };

  return (
    <div style={{ ...stylingProps, overflowY: 'auto', paddingBottom: 3 }}>
      <div style={{
        position: 'absolute',
        bottom: 10,
        left: 0,
        right: 0,
        width: 300,
        display: 'flex',
        justifyContent: 'space-around',
        marginLeft: 'auto',
        marginRight: 'auto',
      }}
      >
        <Fab
          color="error"
          aria-label="close"
          onClick={closePin}
        >
          <CloseIcon fontSize="large" />
        </Fab>

        <Fab
          color="success"
          aria-label="pinners"
          onClick={() => setOpenPinnersDialog(true)}
          disabled={!pinInformation.savedBy.length}
        >
          <StarBorderIcon fontSize="large" />
        </Fab>

        <Fab
          color="primary"
          aria-label="add"
          onClick={handleOpenCommentForm}
          disabled={!authenticated || openCommentForm}
        >
          <AddIcon fontSize="large" />
        </Fab>
      </div>

      <div style={{
        display: 'flex',
        justifyContent: 'space-between',
        alignItems: 'center',
        position: 'sticky',
        top: 0,
        background: 'white',
        marginBottom: 10,
      }}
      >
        <div
          style={{
            display: 'flex',
            justifyContent: 'flex-start',
            alignItems: 'center',
            width: '100%',
            paddingLeft: 10,
            paddingRight: 10,
          }}
        >

          <Avatar
            id="thumbnail"
            alt={pinInformation.imgDescription}
            src={pinInformation.imgLink}
            style={{ width: 56, height: 56, cursor: 'pointer' }}
            onClick={toggleComments}
          />

          <Tags
            commentFormIsOpen={openCommentForm}
            updateTags={updateTags}
            pinInformation={pinInformation}
            closePin={closePin}
          />
        </div>
      </div>
      {
        openCommentForm && (
          <CommentForm
            handleClose={handleCloseCommentForm}
            handleSubmit={submitComment}
          />
        )
      }
      {
        !openCommentForm && !pinInformation.comments.length && (
          <Card
            sx={{ margin: 1 }}
            raised
          >
            <CardContent>
              <Typography variant="h6" color="text.secondary" sx={{ margin: 'auto' }}>
                {authenticated
                  ? `Be the first to write a comment on ${pinInformation.imgDescription}...`
                  : `Login and be the first to write a comment on ${pinInformation.imgDescription}...`}
              </Typography>
            </CardContent>
          </Card>
        )
      }
      {
        !openCommentForm && pinInformation.comments.map(({
          _id, comment, displayName, createdAt,
        }, idx) => (
          <React.Fragment key={_id}>
            <Card
              sx={{ margin: 1 }}
              raised
            >
              <CardContent>
                <div style={{ display: 'flex', alignItems: 'center' }}>
                  <Typography variant="subtitle1" sx={{ color: '#4c62bc', fontWeight: 'bold' }}>
                    {`${displayName}`}
                  </Typography>
                  <Typography color="text.secondary" sx={{ ml: 1, fontSize: 12 }}>
                    {`- ${formatDate(createdAt)}`}
                  </Typography>
                </div>
                <Typography sx={{ mb: 1.5 }} color="text.secondary">
                  {comment}
                </Typography>
              </CardContent>
            </Card>
            {idx === pinInformation.comments.length - 1 && <div style={{ width: '100%', height: 65 }} />}
          </React.Fragment>
        ))
      }
      <PinnersDialog
        open={openPinnersDialog}
        onCloseDialog={() => setOpenPinnersDialog(false)}
        pinnersList={pinInformation.savedBy}
        onClosePin={closePin}
      />
    </div>
  );
}

export default Comments;

Comments.propTypes = {
  pinInformation: PropTypes.objectOf(PropTypes.shape).isRequired,
  stylingProps: PropTypes.objectOf(PropTypes.number).isRequired,
  handleNewComment: PropTypes.func.isRequired,
  authenticated: PropTypes.bool.isRequired,
  toggleComments: PropTypes.func.isRequired,
  closePin: PropTypes.func.isRequired,
  updateTags: PropTypes.func.isRequired,
};
