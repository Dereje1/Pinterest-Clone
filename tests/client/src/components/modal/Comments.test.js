import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Comments from '../../../../../client/src/components/modal/Comments';
import { pinsStub } from '../../../pinsStub';

describe('The comments window', () => {
  let props;
  beforeEach(() => {
    props = {
      stylingProps: { width: 20, height: 30 },
      pinInformation: pinsStub[0],
      handleNewComment: jest.fn(),
      authenticated: true,
      toggleComments: jest.fn(),
      closePin: jest.fn(),
      updateTags: jest.fn(),
    };
  });

  afterEach(() => {
    props = null;
  });

  test('will render with comments', () => {
    const wrapper = shallow(<Comments {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('will render with no comments for authenticated users', () => {
    const updatedProps = { ...props, pinInformation: pinsStub[1] };
    const wrapper = shallow(<Comments {...updatedProps} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('will render with no comments for unauthenticated users', () => {
    const updatedProps = { ...props, pinInformation: pinsStub[1], authenticated: false };
    const wrapper = shallow(<Comments {...updatedProps} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('will open and close the comment form', () => {
    const wrapper = shallow(<Comments {...props} />);
    let fabButton = wrapper.find({ 'aria-label': 'add' });
    let commentForm = wrapper.find('CommentForm');
    // assert closed -> open
    expect(commentForm.isEmptyRender()).toBe(true);
    fabButton.props().onClick();
    commentForm = wrapper.find('CommentForm');
    fabButton = wrapper.find({ 'aria-label': 'add' });
    expect(commentForm.isEmptyRender()).toBe(false);
    expect(fabButton.props().disabled).toBe(true);
    // assert open -> closed
    commentForm.props().handleClose();
    commentForm = wrapper.find('CommentForm');
    expect(commentForm.isEmptyRender()).toBe(true);
  });

  test('will submit a comment', () => {
    const wrapper = shallow(<Comments {...props} />);
    const fabButton = wrapper.find({ 'aria-label': 'add' });
    fabButton.props().onClick();
    const commentForm = wrapper.find('CommentForm');
    commentForm.props().handleSubmit('a new comment');
    expect(props.handleNewComment).toHaveBeenCalledWith('a new comment');
  });

  test('will not submit a comment if empty', () => {
    const wrapper = shallow(<Comments {...props} />);
    const fabButton = wrapper.find({ 'aria-label': 'add' });
    fabButton.props().onClick();
    const commentForm = wrapper.find('CommentForm');
    commentForm.props().handleSubmit('      ');
    expect(props.handleNewComment).not.toHaveBeenCalled();
  });

  test('will open and close the pinners dialog', () => {
    const wrapper = shallow(<Comments {...props} />);
    const pinnersButton = wrapper.find({ 'aria-label': 'pinners' });
    let pinnersDialog = wrapper.find('PinnersDialog');
    // assert closed -> open
    expect(pinnersDialog.props().open).toBe(false);
    pinnersButton.props().onClick();
    pinnersDialog = wrapper.find('PinnersDialog');
    expect(pinnersDialog.props().open).toBe(true);
    // assert open -> closed
    pinnersDialog.props().onCloseDialog();
    pinnersDialog = wrapper.find('PinnersDialog');
    expect(pinnersDialog.props().open).toBe(false);
  });

  test('will not render pinners button if pin has no pinners', () => {
    const updatedProps = {
      ...props,
      pinInformation: {
        ...props.pinInformation,
        savedBy: [],
      },
    };
    const wrapper = shallow(<Comments {...updatedProps} />);
    const pinnersButton = wrapper.find('ForwardRef(Button)');
    expect(pinnersButton.isEmptyRender()).toBe(true);
  });

  test('will toggle the comments window on thumbnail img click', () => {
    const wrapper = shallow(<Comments {...props} />);
    const thumbnailDiv = wrapper.find({ id: 'thumbnail' });
    thumbnailDiv.props().onClick();
    thumbnailDiv.props().onKeyDown();
    expect(props.toggleComments).toHaveBeenCalled();
  });
});
