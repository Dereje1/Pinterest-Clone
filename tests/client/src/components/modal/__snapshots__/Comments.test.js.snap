// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The comments window will render with comments 1`] = `
<div
  style={
    Object {
      "height": 30,
      "overflowY": "auto",
      "paddingBottom": 3,
      "width": 20,
    }
  }
>
  <ForwardRef(Fab)
    aria-label="add"
    color="primary"
    disabled={false}
    onClick={[Function]}
    onMouseDown={[Function]}
    sx={
      Object {
        "bottom": 10,
        "position": "absolute",
        "right": 10,
      }
    }
  >
    <Memo(ForwardRef(AddIcon)) />
  </ForwardRef(Fab)>
  <div
    style={
      Object {
        "alignItems": "center",
        "background": "white",
        "display": "flex",
        "justifyContent": "space-between",
        "marginBottom": 10,
        "position": "sticky",
        "top": 0,
      }
    }
  >
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "justifyContent": "space-between",
          "paddingLeft": 10,
          "paddingRight": 10,
          "width": "100%",
        }
      }
    >
      <div
        id="thumbnail"
        onClick={[MockFunction]}
        onKeyDown={[Function]}
        role="button"
        style={
          Object {
            "cursor": "pointer",
          }
        }
        tabIndex="0"
      >
        <img
          alt="imgDescription id-1"
          src="imgLink id-1"
          style={
            Object {
              "height": 56,
              "objectFit": "scale-down",
              "width": 56,
            }
          }
        />
      </div>
      <ForwardRef(Button)
        onClick={[Function]}
        variant="text"
      >
        Pinners
      </ForwardRef(Button)>
      <ForwardRef(IconButton)
        onClick={[MockFunction]}
      >
        <Memo(ForwardRef(CloseIcon)) />
      </ForwardRef(IconButton)>
    </div>
  </div>
  <ForwardRef(Card)
    raised={true}
    sx={
      Object {
        "margin": 1,
      }
    }
  >
    <ForwardRef(CardContent)>
      <div
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
          }
        }
      >
        <ForwardRef(Typography)
          sx={
            Object {
              "color": "#4c62bc",
              "fontWeight": "bold",
            }
          }
          variant="subtitle1"
        >
          tester-displayName
        </ForwardRef(Typography)>
        <ForwardRef(Typography)
          color="text.secondary"
          sx={
            Object {
              "fontSize": 12,
              "ml": 1,
            }
          }
        >
          - 28 Jul 2021
        </ForwardRef(Typography)>
      </div>
      <ForwardRef(Typography)
        color="text.secondary"
        sx={
          Object {
            "mb": 1.5,
          }
        }
      >
        Test comment
      </ForwardRef(Typography)>
    </ForwardRef(CardContent)>
  </ForwardRef(Card)>
  <PinnersDialog
    onCloseDialog={[Function]}
    onClosePin={[MockFunction]}
    open={false}
    pinnersList={
      Array [
        Object {
          "name": "savedBy - id-1",
          "service": "google",
          "userId": "1",
        },
        Object {
          "name": "savedBy - id-1",
          "service": "github",
          "userId": "1",
        },
      ]
    }
  />
</div>
`;

exports[`The comments window will render with no comments for authenticated users 1`] = `
<div
  style={
    Object {
      "height": 30,
      "overflowY": "auto",
      "paddingBottom": 3,
      "width": 20,
    }
  }
>
  <ForwardRef(Fab)
    aria-label="add"
    color="primary"
    disabled={false}
    onClick={[Function]}
    onMouseDown={[Function]}
    sx={
      Object {
        "bottom": 10,
        "position": "absolute",
        "right": 10,
      }
    }
  >
    <Memo(ForwardRef(AddIcon)) />
  </ForwardRef(Fab)>
  <div
    style={
      Object {
        "alignItems": "center",
        "background": "white",
        "display": "flex",
        "justifyContent": "space-between",
        "marginBottom": 10,
        "position": "sticky",
        "top": 0,
      }
    }
  >
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "justifyContent": "space-between",
          "paddingLeft": 10,
          "paddingRight": 10,
          "width": "100%",
        }
      }
    >
      <div
        id="thumbnail"
        onClick={[MockFunction]}
        onKeyDown={[Function]}
        role="button"
        style={
          Object {
            "cursor": "pointer",
          }
        }
        tabIndex="0"
      >
        <img
          alt="imgDescription id-2"
          src="imgLink id-2"
          style={
            Object {
              "height": 56,
              "objectFit": "scale-down",
              "width": 56,
            }
          }
        />
      </div>
      <ForwardRef(Button)
        onClick={[Function]}
        variant="text"
      >
        Pinners
      </ForwardRef(Button)>
      <ForwardRef(IconButton)
        onClick={[MockFunction]}
      >
        <Memo(ForwardRef(CloseIcon)) />
      </ForwardRef(IconButton)>
    </div>
  </div>
  <ForwardRef(Card)
    raised={true}
    sx={
      Object {
        "margin": 1,
      }
    }
  >
    <ForwardRef(CardContent)>
      <ForwardRef(Typography)
        color="text.secondary"
        sx={
          Object {
            "margin": "auto",
          }
        }
        variant="h6"
      >
        Be the first to write a comment on imgDescription id-2...
      </ForwardRef(Typography)>
    </ForwardRef(CardContent)>
  </ForwardRef(Card)>
  <PinnersDialog
    onCloseDialog={[Function]}
    onClosePin={[MockFunction]}
    open={false}
    pinnersList={
      Array [
        Object {
          "name": "savedBy - id-2",
          "service": "google",
          "userId": "2",
        },
        Object {
          "name": "savedBy - id-2git",
          "service": "github",
          "userId": "2",
        },
      ]
    }
  />
</div>
`;

exports[`The comments window will render with no comments for unauthenticated users 1`] = `
<div
  style={
    Object {
      "height": 30,
      "overflowY": "auto",
      "paddingBottom": 3,
      "width": 20,
    }
  }
>
  <ForwardRef(Fab)
    aria-label="add"
    color="primary"
    disabled={true}
    onClick={[Function]}
    onMouseDown={[Function]}
    sx={
      Object {
        "bottom": 10,
        "position": "absolute",
        "right": 10,
      }
    }
  >
    <Memo(ForwardRef(AddIcon)) />
  </ForwardRef(Fab)>
  <div
    style={
      Object {
        "alignItems": "center",
        "background": "white",
        "display": "flex",
        "justifyContent": "space-between",
        "marginBottom": 10,
        "position": "sticky",
        "top": 0,
      }
    }
  >
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "justifyContent": "space-between",
          "paddingLeft": 10,
          "paddingRight": 10,
          "width": "100%",
        }
      }
    >
      <div
        id="thumbnail"
        onClick={[MockFunction]}
        onKeyDown={[Function]}
        role="button"
        style={
          Object {
            "cursor": "pointer",
          }
        }
        tabIndex="0"
      >
        <img
          alt="imgDescription id-2"
          src="imgLink id-2"
          style={
            Object {
              "height": 56,
              "objectFit": "scale-down",
              "width": 56,
            }
          }
        />
      </div>
      <ForwardRef(Button)
        onClick={[Function]}
        variant="text"
      >
        Pinners
      </ForwardRef(Button)>
      <ForwardRef(IconButton)
        onClick={[MockFunction]}
      >
        <Memo(ForwardRef(CloseIcon)) />
      </ForwardRef(IconButton)>
    </div>
  </div>
  <ForwardRef(Card)
    raised={true}
    sx={
      Object {
        "margin": 1,
      }
    }
  >
    <ForwardRef(CardContent)>
      <ForwardRef(Typography)
        color="text.secondary"
        sx={
          Object {
            "margin": "auto",
          }
        }
        variant="h6"
      >
        Login and be the first to write a comment on imgDescription id-2...
      </ForwardRef(Typography)>
    </ForwardRef(CardContent)>
  </ForwardRef(Card)>
  <PinnersDialog
    onCloseDialog={[Function]}
    onClosePin={[MockFunction]}
    open={false}
    pinnersList={
      Array [
        Object {
          "name": "savedBy - id-2",
          "service": "google",
          "userId": "2",
        },
        Object {
          "name": "savedBy - id-2git",
          "service": "github",
          "userId": "2",
        },
      ]
    }
  />
</div>
`;
